src/layout/rect/iter.rs

```diff
@@ -48,6 +50,7 @@ impl Columns {
         Self {
             rect,
             current_column: rect.x,
+            current_column_back: rect.x + rect.width,
```

Comment:
```suggestion
            current_column_back: rect.right(),
```

That said, I found it a little confusing that the current column is the next column to return, but the back column is the the next_back + 1 column. Does it make sense here to -1 on this and then use > instead of >= in the checks below?

e.g. for a 3 column area, this will currently be:
current_column = 0
current_column_back = 3 (should be 2)

Reply:
Changed to use `rect.right()`.

Things seem to work out badly when `current_column_back` points to the (and not past the) final index, mainly with the terminate case for `next_back()`.  I note that `right()` is defined as "the first coordinate outside of the `Rect`" which makes sense to me and seems consistent with this.  I tried to get it to work anyway but couldn't, maybe i'm just being stupid, feel free to have a fiddle.

Reply:
Ah, you're right. The thing I was missing was that the indexes need to record when the iterators cross explicitly. A good simple case to test this is having a single item. If the indexes are both 0, then there's nothing to disambiguate whether the iterator has been exhausted already or if the iterator is in its initial state. Having them as 0, 1 makes this possible.

I wonder if there's problems around u16::MAX for this.

Reply:
> I wonder if there's problems around u16::MAX for this.

I added a couple of boundary condition tests and I think it is correct.

```rust
/// We allow a total of `65536` columns in the range `(0..=65535)`.  In this test we iterate
/// forward and skip the first `65534` columns, and expect the next column to be `65535` and
/// the subsequent columns to be `None`.
#[test]
fn columns_max() {
    let rect = Rect::new(0, 0, u16::MAX, 1);
    let mut columns = Columns::new(rect).skip(usize::from(u16::MAX - 1));
    assert_eq!(columns.next(), Some(Rect::new(u16::MAX - 1, 0, 1, 1)));
    assert_eq!(columns.next(), None);
}

/// We allow a total of `65536` columns in the range `(0..=65535)`.  In this test we iterate
/// backward and skip the last `65534` columns, and expect the next column to be `0` and the
/// subsequent columns to be `None`.
#[test]
fn columns_min() {
    let rect = Rect::new(0, 0, u16::MAX, 1);
    let mut columns = Columns::new(rect).rev().skip(usize::from(u16::MAX - 1));
    assert_eq!(columns.next(), Some(Rect::new(0, 0, 1, 1)));
    assert_eq!(columns.next(), None);
}
```

Reply:
thanks :)
Usually these things are not a huge problem, but someone somewhere is likely to try to break some random tui with this sort of thing.